File: playground/android/app/src/main/java/com/amap/agenuiplayground/stability/RealisticScenarioEngine.java
package com.amap.agenuiplayground.stability;

import android.app.Activity;
import android.os.Handler;
import android.os.Looper;
import android.util.Log;

import com.amap.agenui.render.surface.ISurfaceManagerListener;
import com.amap.agenui.render.surface.Surface;
import com.amap.agenui.render.surface.SurfaceManager;
import com.amap.agenui.render.surface.SurfaceSize;

import org.json.JSONArray;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Async engine that executes realistic scenarios with visible UI rendering.
 * Each scenario loads a fixture JSON with timed steps, renders surfaces full-screen,
 * and reports results via listener callback.
 */
public class RealisticScenarioEngine implements ISurfaceManagerListener {
    private static final String TAG = "RealisticEngine";
    private static final String FIXTURES_DIR = "stability_fixtures/realistic_scenarios";

    // Fixture step model
    private static class FixtureStep {
        final String action;
        final int delayMs;
        final String message;
        final String raw;

        FixtureStep(String action, int delayMs, String message, String raw) {
            this.action = action;
            this.delayMs = delayMs;
            this.message = message;
            this.raw = raw;
        }
    }

    // Listener interface for Activity integration
    public interface Listener {
        void onSurfaceCreated(Surface surface);
        void onSurfaceDeleted(Surface surface);
        void onRoundCompleted(String fixture, String error);
    }

    // Scenario to fixture name mapping
    private static final Map<StabilityScenarioEngine.Scenario, String> FIXTURE_MAP = new HashMap<>();
    static {
        FIXTURE_MAP.put(StabilityScenarioEngine.Scenario.REALISTIC_ARTICLE_STREAM, "article_stream");
        FIXTURE_MAP.put(StabilityScenarioEngine.Scenario.REALISTIC_MULTI_CARD, "multi_card");
        FIXTURE_MAP.put(StabilityScenarioEngine.Scenario.REALISTIC_FORM_FILL, "form_fill");
        FIXTURE_MAP.put(StabilityScenarioEngine.Scenario.REALISTIC_CHART_REFRESH, "chart_refresh");
        FIXTURE_MAP.put(StabilityScenarioEngine.Scenario.REALISTIC_LONG_LIST, "long_list");
        FIXTURE_MAP.put(StabilityScenarioEngine.Scenario.REALISTIC_PAGE_SWITCH, "page_switch");
        FIXTURE_MAP.put(StabilityScenarioEngine.Scenario.REALISTIC_TAB_NAVIGATION, "tab_navigation");
        FIXTURE_MAP.put(StabilityScenarioEngine.Scenario.REALISTIC_LOTTIE_CAROUSEL, "lottie_carousel");
        FIXTURE_MAP.put(StabilityScenarioEngine.Scenario.REALISTIC_MIXED_DASHBOARD, "mixed_dashboard");
        FIXTURE_MAP.put(StabilityScenarioEngine.Scenario.REALISTIC_ERROR_RECOVERY, "error_recovery");
    }

    private final Activity activity;
    private final Handler handler;
    private final StabilityScenarioEngine stressEngine;

    // Execution state
    private SurfaceManager surfaceManager;
    private final Map<String, Surface> activeSurfaces = new HashMap<>();
    private boolean isExecuting = false;
    private boolean cancelled = false;
    private String currentFixtureName;
    private List<FixtureStep> currentSteps;
    private int stepIndex;
    private Listener listener;

    public RealisticScenarioEngine(Activity activity, StabilityScenarioEngine stressEngine) {
        this.activity = activity;
        this.handler = new Handler(Looper.getMainLooper());
        this.stressEngine = stressEngine;
    }

    /**
     * Execute a realistic scenario round asynchronously.
     * Must be called on the main thread.
     */
    public void executeRound(StabilityScenarioEngine.Scenario scenario, Listener listener) {
        if (isExecuting) {
            listener.onRoundCompleted(null, "Engine busy");
            return;
        }
        if (!scenario.isRealistic()) {
            listener.onRoundCompleted(null, "Not a realistic scenario");
            return;
        }

        String fixtureName = FIXTURE_MAP.get(scenario);
        if (fixtureName == null) {
            listener.onRoundCompleted(null, "No fixture for " + scenario.name());
            return;
        }

        List<FixtureStep> steps = loadFixture(fixtureName);
        if (steps == null) {
            listener.onRoundCompleted(fixtureName, "Failed to load fixture");
            return;
        }

        isExecuting = true;
        cancelled = false;
        this.listener = listener;
        this.currentFixtureName = fixtureName;
        this.currentSteps = steps;
        this.stepIndex = 0;

        // Create fresh SurfaceManager
        surfaceManager = new SurfaceManager(activity);
        surfaceManager.addListener(this);
        surfaceManager.beginTextStream();

        // Start executing steps
        executeNextStep();
    }

    /**
     * Cancel current execution.
     */
    public void cancel() {
        if (!isExecuting) return;
        cancelled = true;
        finishRound("Cancelled");
    }

    /**
     * Tear down any in-progress state.
     */
    public void teardown() {
        cancelled = true;
        handler.removeCallbacksAndMessages(null);
        cleanupSurfaces();
        if (surfaceManager != null) {
            surfaceManager.removeListener(this);
            surfaceManager.destroy();
            surfaceManager = null;
        }
        isExecuting = false;
        listener = null;
    }

    // MARK: - Step Execution

    private void executeNextStep() {
        if (cancelled) return;
        if (stepIndex >= currentSteps.size()) {
            // All steps done
            if (surfaceManager != null) {
                surfaceManager.endTextStream();
            }
            handler.postDelayed(() -> finishRound(null), 200);
            return;
        }

        FixtureStep step = currentSteps.get(stepIndex);
        stepIndex++;

        int delay = Math.max(step.delayMs, 0);
        handler.postDelayed(() -> {
            if (cancelled) return;
            performStep(step);
            executeNextStep();
        }, delay);
    }

    private void performStep(FixtureStep step) {
        if (surfaceManager == null) return;

        switch (step.action) {
            case "createSurface":
            case "updateComponents":
            case "updateDataModel":
            case "deleteSurface":
                if (step.message != null) {
                    surfaceManager.receiveTextChunk(step.message);
                }
                break;

            case "rawChunk":
                if (step.raw != null) {
                    surfaceManager.receiveTextChunk(step.raw);
                }
                break;

            case "beginNewStream":
                surfaceManager.endTextStream();
                surfaceManager.beginTextStream();
                break;

            default:
                if (step.message != null) {
                    surfaceManager.receiveTextChunk(step.message);
                }
                break;
        }
    }

    // MARK: - ISurfaceManagerListener

    @Override
    public void onCreateSurface(Surface surface) {
        activeSurfaces.put(surface.getSurfaceId(), surface);
        if (listener != null) {
            listener.onSurfaceCreated(surface);
        }
    }

    @Override
    public void onDeleteSurface(Surface surface) {
        activeSurfaces.remove(surface.getSurfaceId());
        if (listener != null) {
            listener.onSurfaceDeleted(surface);
        }
    }

    @Override
    public void onReceiveActionEvent(String event) {
    }

    @Override
    public void onRootComponentUpdate(Surface surface, Map<String, String> props) {
    }

    @Override
    public void onError(Surface surface, int code, String message) {
    }

    @Override
    public void onBlankCheckResult(Surface surface, boolean isBlank) {
    }

    @Override
    public void onComponentAppeared(Surface surface, String parentComponentId, String parentType, Map<String, Object> properties) {
    }

    @Override
    public SurfaceSize surfaceSize(String surfaceId) {
        return null;
    }

    // MARK: - Finish & Cleanup

    private void finishRound(String error) {
        handler.removeCallbacksAndMessages(null);
        cleanupSurfaces();
        if (surfaceManager != null) {
            surfaceManager.removeListener(this);
            surfaceManager.destroy();
            surfaceManager = null;
        }
        isExecuting = false;

        String fixture = currentFixtureName;
        Listener cb = listener;
        currentFixtureName = null;
        currentSteps = null;
        stepIndex = 0;
        listener = null;

        if (cb != null) {
            cb.onRoundCompleted(fixture, error);
        }
    }

    private void cleanupSurfaces() {
        if (listener != null) {
            for (Surface surface : activeSurfaces.values()) {
                listener.onSurfaceDeleted(surface);
            }
        }
        activeSurfaces.clear();
    }

    // MARK: - Fixture Loading

    private List<FixtureStep> loadFixture(String name) {
        String path = FIXTURES_DIR + "/" + name + ".json";
        String json = stressEngine.loadAssetFile(path);
        if (json == null) return null;

        try {
            JSONObject root = new JSONObject(json);
            JSONArray stepsArray = root.getJSONArray("steps");

            List<FixtureStep> steps = new ArrayList<>();
            for (int i = 0; i < stepsArray.length(); i++) {
                JSONObject stepObj = stepsArray.getJSONObject(i);
                String action = stepObj.optString("action", "unknown");
                int delayMs = stepObj.optInt("delay_ms", 0);

                String message = null;
                if (stepObj.has("message")) {
                    message = stepObj.getJSONObject("message").toString();
                }

                String raw = stepObj.optString("raw", null);

                steps.add(new FixtureStep(action, delayMs, message, raw));
            }
            return steps;
        } catch (Exception e) {
            Log.e(TAG, "Failed to parse fixture: " + name, e);
            return null;
        }
    }
}

File: playground/ios/Playground/Playground for AGenUI/Stability/RealisticScenarioEngine.swift
import Foundation
import UIKit
import AGenUI

/// Async engine that executes realistic scenarios with visible UI rendering.
/// Each scenario loads a fixture JSON with timed steps, renders surfaces full-screen,
/// and reports results via completion handler.
class RealisticScenarioEngine: NSObject, SurfaceManagerListener {

    // MARK: - Types

    private struct FixtureStep {
        let action: String
        let delayMs: Int
        let message: String?   // serialized JSON for normal actions
        let raw: String?       // raw text for error injection
    }

    private struct Fixture {
        let description: String
        let surfaceId: String
        let isMultiSurface: Bool
        let steps: [FixtureStep]
    }

    // MARK: - Callbacks

    /// Called when a surface view is created and should be added to the view hierarchy.
    var onSurfaceCreated: ((Surface) -> Void)?

    /// Called when a surface is deleted and its view should be removed.
    var onSurfaceDeleted: ((Surface) -> Void)?

    /// Called when all steps have finished for the current round.
    private var roundCompletion: ((ScenarioRoundResult) -> Void)?

    // MARK: - State

    private var surfaceManager: SurfaceManager?
    private var activeSurfaces: [String: Surface] = [:]
    private var isExecuting = false
    private var currentFixtureName: String?
    private var stepIndex = 0
    private var currentSteps: [FixtureStep] = []
    private var executionCancelled = false

    // MARK: - Fixture Mapping

    private static let scenarioFixtureMap: [StabilityScenario: String] = [
        .realisticArticleStream: "article_stream",
        .realisticMultiCard: "multi_card",
        .realisticFormFill: "form_fill",
        .realisticChartRefresh: "chart_refresh",
        .realisticLongList: "long_list",
        .realisticPageSwitch: "page_switch",
        .realisticTabNavigation: "tab_navigation",
        .realisticLottieCarousel: "lottie_carousel",
        .realisticMixedDashboard: "mixed_dashboard",
        .realisticErrorRecovery: "error_recovery",
    ]

    // MARK: - Public API

    /// Execute a realistic scenario round asynchronously.
    /// Must be called on the main thread.
    /// - Parameters:
    ///   - scenario: The realistic scenario to execute.
    ///   - containerWidth: Width constraint for surface rendering.
    ///   - completion: Called on the main thread when the round finishes.
    func executeRound(_ scenario: StabilityScenario, containerWidth: CGFloat, completion: @escaping (ScenarioRoundResult) -> Void) {
        guard !isExecuting else {
            completion(ScenarioRoundResult(fixture: nil, error: "Engine busy"))
            return
        }
        guard scenario.isRealistic else {
            completion(ScenarioRoundResult(fixture: nil, error: "Not a realistic scenario"))
            return
        }

        guard let fixtureName = Self.scenarioFixtureMap[scenario] else {
            completion(ScenarioRoundResult(fixture: nil, error: "No fixture for \(scenario.rawValue)"))
            return
        }

        guard let fixture = loadFixture(name: fixtureName) else {
            completion(ScenarioRoundResult(fixture: fixtureName, error: "Failed to load fixture"))
            return
        }

        isExecuting = true
        executionCancelled = false
        currentFixtureName = fixtureName
        currentSteps = fixture.steps
        stepIndex = 0
        roundCompletion = completion

        // Create a fresh SurfaceManager for this round
        let sm = SurfaceManager()
        sm.addListener(self)
        surfaceManager = sm
        sm.beginTextStream()

        // Start executing steps
        executeNextStep(containerWidth: containerWidth)
    }

    /// Cancel current execution. The completion handler will be called with a cancellation error.
    func cancel() {
        guard isExecuting else { return }
        executionCancelled = true
        finishRound(error: "Cancelled")
    }

    /// Tear down any in-progress state (e.g., when the VC disappears).
    func teardown() {
        executionCancelled = true
        cleanupSurfaces()
        surfaceManager?.endTextStream()
        surfaceManager?.removeAllListeners()
        surfaceManager = nil
        isExecuting = false
        roundCompletion = nil
    }

    // MARK: - Step Execution

    private func executeNextStep(containerWidth: CGFloat) {
        guard !executionCancelled else { return }
        guard stepIndex < currentSteps.count else {
            // All steps done — end stream and finish
            surfaceManager?.endTextStream()
            // Allow a brief delay for final layout to settle
            DispatchQueue.main.asyncAfter(deadline: .now() + .milliseconds(200)) { [weak self] in
                self?.finishRound(error: nil)
            }
            return
        }

        let step = currentSteps[stepIndex]
        stepIndex += 1

        let delay = DispatchTimeInterval.milliseconds(max(step.delayMs, 0))
        DispatchQueue.main.asyncAfter(deadline: .now() + delay) { [weak self] in
            guard let self = self, !self.executionCancelled else { return }
            self.performStep(step, containerWidth: containerWidth)
            self.executeNextStep(containerWidth: containerWidth)
        }
    }

    private func performStep(_ step: FixtureStep, containerWidth: CGFloat) {
        guard let sm = surfaceManager else { return }

        switch step.action {
        case "createSurface", "updateComponents", "updateDataModel", "deleteSurface":
            if let message = step.message {
                sm.receiveTextChunk(message)
            }

        case "rawChunk":
            // Inject raw (potentially malformed) data
            if let raw = step.raw {
                sm.receiveTextChunk(raw)
            }

        case "beginNewStream":
            // Simulate interrupt: end current stream, begin a new one
            sm.endTextStream()
            sm.beginTextStream()

        default:
            // Unknown action — treat message as raw chunk if available
            if let message = step.message {
                sm.receiveTextChunk(message)
            }
        }
    }

    // MARK: - SurfaceManagerListener

    func onCreateSurface(_ surface: Surface) {
        let screenWidth = UIScreen.main.bounds.width
        surface.updateSize(width: screenWidth, height: .infinity)
        activeSurfaces[surface.surfaceId] = surface

        surface.onLayoutChanged = { [weak surface] in
            guard let s = surface else { return }
            // Update view frame for scroll sizing
            let height = s.view.frame.size.height
            if height > 0 {
                s.view.frame = CGRect(x: 0, y: s.view.frame.origin.y,
                                      width: screenWidth, height: height)
            }
        }

        onSurfaceCreated?(surface)
    }

    func onDeleteSurface(_ surface: Surface) {
        activeSurfaces.removeValue(forKey: surface.surfaceId)
        onSurfaceDeleted?(surface)
    }

    // MARK: - Finish & Cleanup

    private func finishRound(error: String?) {
        cleanupSurfaces()
        surfaceManager?.removeAllListeners()
        surfaceManager = nil
        isExecuting = false

        let result = ScenarioRoundResult(fixture: currentFixtureName, error: error)
        let completion = roundCompletion
        roundCompletion = nil
        currentFixtureName = nil
        currentSteps = []
        stepIndex = 0

        completion?(result)
    }

    private func cleanupSurfaces() {
        for (_, surface) in activeSurfaces {
            onSurfaceDeleted?(surface)
        }
        activeSurfaces.removeAll()
    }

    // MARK: - Fixture Loading

    private func loadFixture(name: String) -> Fixture? {
        guard let path = Bundle.main.path(forResource: name, ofType: "json",
                                          inDirectory: "stability_fixtures/realistic_scenarios"),
              let data = FileManager.default.contents(atPath: path) else {
            return nil
        }

        guard let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            return nil
        }

        let description = json["description"] as? String ?? name
        let surfaceId = json["surfaceId"] as? String ?? "realistic-\(name)"
        let isMultiSurface = json["multi_surface"] as? Bool ?? false

        guard let stepsArray = json["steps"] as? [[String: Any]] else {
            return nil
        }

        var steps: [FixtureStep] = []
        for stepDict in stepsArray {
            let action = stepDict["action"] as? String ?? "unknown"
            let delayMs = stepDict["delay_ms"] as? Int ?? 0

            var messageStr: String?
            if let messageObj = stepDict["message"] {
                if let msgData = try? JSONSerialization.data(withJSONObject: messageObj),
                   let msgStr = String(data: msgData, encoding: .utf8) {
                    messageStr = msgStr
                }
            }

            let raw = stepDict["raw"] as? String

            steps.append(FixtureStep(action: action, delayMs: delayMs, message: messageStr, raw: raw))
        }

        return Fixture(description: description, surfaceId: surfaceId,
                       isMultiSurface: isMultiSurface, steps: steps)
    }
}
